diff --git a/src/lib_jit.c b/src/lib_jit.c
index fd8e585b..f006f2ce 100644
--- a/src/lib_jit.c
+++ b/src/lib_jit.c
@@ -701,7 +701,8 @@ static uint32_t jit_cpudetect(void)
     if (x) flags |= JIT_F_MIPSXXR2;  /* Either 0x80000000 (R2) or 0 (R1). */
   }
 #endif
-
+#elif LJ_TARGET_S390X
+  /* No optional CPU features to detect (for now). */
 #else
 #error "Missing CPU detection for this architecture"
 #endif
