diff --git a/src/lj_mcode.c b/src/lj_mcode.c
index 2b8ac2df..54f8123d 100644
--- a/src/lj_mcode.c
+++ b/src/lj_mcode.c
@@ -38,6 +38,7 @@
 void sys_icache_invalidate(void *start, size_t len);
 #endif
 
+#ifndef __MVS__
 /* Synchronize data/instruction cache. */
 void lj_mcode_sync(void *start, void *end)
 {
@@ -59,6 +60,7 @@ void lj_mcode_sync(void *start, void *end)
 #endif
 }
 
+#endif
 #endif
 
 #if LJ_HASJIT
